### ISSUE TITLE
Type-safe `unittest.mock.patch`

### ISSUE THREAD

**Issue by not-my-profile** [microsoft/pyright#7108]
For example:

```python
# bar.py
def foo() -> int:
    return 3
```

```python
# foo.py
import unittest.mock
import bar

with unittest.mock.patch('bar.foo') as foo:
    foo.return_value = None # no type error
    print(bar.foo())
```

I would like this `return_value` assignment to result in a type error since the new return value is incompatible with the declared type of the `bar.foo` method. I know that this currently isn't possible since Pyright doesn't have special knowledge about the `unittest.mock` semantics. Nonetheless I am working with a large codebase that uses a lot of `mock.patch` to patch return values in unit tests, so being able to statically type check this patching would greatly improve the type safety in the tests.

**Comment by erictraut** [comment]
The type definition for `mock.patch` is declared in the typeshed stub `mock.pyi`, and pyright is correctly applying the typing rules according to the current definitions in this stub.

If you would like to see the definition changed, please post an issue to the typeshed project. The typeshed maintainers can provide more insights about why it is declared the way it is. 

In general, monkey-patching (which is effectively what `unittest.mock.patch` is doing) is not possible to statically type check because it violates some of the basic assumptions of static type checking.
